File: 35.py
# 35. 搜索插入位置
class Solution(object):
    def searchInsert(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        if not nums:
            return 0
        length = len(nums)
        low = 0
        high = length - 1
        while low <= high:
            mid = (low + high) / 2
            if nums[mid] < target:
                low = mid + 1
            elif nums[mid] > target:
                high = mid - 1
            else:
                return mid
        return low

File: 35.go
package main

import "fmt"

func main() {
	fmt.Println(searchInsert([]int{1, 3, 5, 6}, 5))
	fmt.Println(searchInsert([]int{1, 3, 5, 6}, 2))
	fmt.Println(searchInsert([]int{1, 3, 5, 6}, 7))
	fmt.Println(searchInsert([]int{1, 3, 5, 6}, 0))
	fmt.Println(searchInsert([]int{1}, 0))
	fmt.Println(searchInsert([]int{1, 3}, 2))
}

func searchInsert(nums []int, target int) int {
	i, j := 0, len(nums) - 1
	ans := len(nums)
	for i <= j {
		mid := (j-i)/2 + i
		n := nums[mid]
		if n >= target {
			ans = mid
			j = mid - 1
		} else {
			i = mid + 1
		}
	}
	return ans
}
